flink-autoscaler/src/main/java/org/apache/flink/autoscaler/tuning/MemoryTuning.java

```diff
@@ -277,6 +286,32 @@ private static MemorySize adjustNetworkMemory(
         return new MemorySize(memBudget.budget(maxNetworkMemory));
     }
 
+    /**
+     * Calculate how many network segment current vertex needs.
+     *
+     * @param currentVertexParallelism The parallelism of current vertex.
+     * @param otherVertexParallelism The parallelism of other vertex.
+     */
+    @VisibleForTesting
+    static int calculateNetworkSegmentNumber(
+            int currentVertexParallelism,
+            int otherVertexParallelism,
+            String shipStrategy,
+            int buffersPerChannel,
+            int floatingBuffers) {
+        // TODO When the parallelism is changed via the rescale api, the FORWARD may be changed to
+        // RESCALE. This logic may needs to be updated after FLINK-33123.
+        if (currentVertexParallelism == otherVertexParallelism && "FORWARD".equals(shipStrategy)) {
+            return buffersPerChannel + floatingBuffers;
+        } else if ("FORWARD".equals(shipStrategy) || "RESCALE".equals(shipStrategy)) {
```

Comment:
The shipStragegy in the json plan comes from https://github.com/apache/flink/blob/6c8f3a0799c609d8076f782e5334e389e4d92eee/flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java#L1433. 

This is a bit concerning because any custom partitioner which has a `toString()` implementation which returns FORWARD, would run into this code branch. Although unlikely, that would break the network memory code.

Reply:
It comes from `org.apache.flink.streaming.runtime.partitioner.StreamPartitioner#toString` instead of `org.apache.flink.api.common.functions.Partitioner`.

IIUC, users implement `org.apache.flink.api.common.functions.Partitioner`. And `CustomPartitionerWrapper` wraps the `Partitioner`. The `CustomPartitionerWrapper#toString` is `CUSTOM`.

I have wrote a demo, and overwrite the toString of custom `Partitioner`. It still shows `CUSTOM`.

Reply:
Got it, that makes sense. So the toString() can't be overwritten by users to shadow the built-in ones.